RFC-0004-sparse-fill-value/README.md

```diff
@@ -0,0 +1,615 @@
+    `sparseDims` when specified: `self.shape[:sparseDims]` must be
+    equal to `<constructed sparse tensor>.fill_value().shape`.
+
+12. Sparse tensors with defined fill value have intrinsic constraints
+    between all the unspecified tensor elements (these are always
+    equal) that must be taken into account when implementing Autograd
+    backward methods for functions that receive sparse tensors as
+    inputs.
+
+
+### Future extensions and existing issues
+
+Introducing the fill value feature according to this proposal does not
+require addressing the following extensions and issues. These are
+given here as suggestions to clean up the PyTorch sparse tensor
+support in general.
+
+13. For the Graph domain, the indefinite fill value can be specified as a
+    tensor with zero dimension(s) that satisfies all the relations
+    listed above except point 5. Invalidation of the point 5 will
+    provide a consistent way to differentiate between defined and
+    indefinite fill values.
+
+14. The acronym NNZ means the "Number of nonzeros" in a sparse
+    tensor. In PyTorch, this acronym is used in several places:
+
+    - in the implementations of sparse tensor and related
+      functionality,
+    - in the `repr` output of COO sparse tensor, see
+      [here](https://pytorch.org/docs/master/generated/torch.sparse_coo_tensor.html#torch.sparse_coo_tensor),
+    - as a private method `_nnz()`, see
+      [here](https://pytorch.org/docs/master/sparse.html?highlight=nnz#torch.sparse.FloatTensor._nnz),
+    - as a optional keyword argument `check_sparse_nnz` in
+      [torch.autograd.gradcheck](https://pytorch.org/docs/master/autograd.html#numerical-gradient-checking).
+
+    The acronym NNZ is misused in PyTorch:
+
+    - `nnz` holds the value of the "Number of Specified Elements"
+      (NSE) in a sparse tensor
+    - `nnz` is not always equal to the number of zeros in the sparse
+      tensor, for instance, the `values` of the sparse tensor in COO
+      format may contain zero values that are not accounted in `nnz`
+
+    With the introduction of nonzero fill values, the misuse of
+    acronym NNZ will get worse because with nonzero fill value the
+    sparse tensor may have no zero elements, e.g. `torch.full((10,
+    10), 1.0, layout=torch.sparse_coo)` for which `nnz` would be `0`.
+
+    For the reasons above, this proposal recommends
+
+    - renaming NNZ to NSE,
+    - deprecating the `_nnz()` method in favor of `_nse()` method,
+    - removing the `_nnz()` method starting from PyTorch 2.0.
+
+    Alternative: Do nothing.  This is the (undocumented) approach
+    taken in Wolfram Language where [one can use "NonzeroValues" to
+    determine the number of specified
+    elements](https://community.wolfram.com/groups/-/m/t/1168496) even
+    when the fill value specified is nonzero.
+
+15. The `torch` namespace functions `arange`, `range`, `linspace`, and
+    `logspace` have `layout` argument that is not needed.
```

Comment:
Although I don't have concrete evidence for it, I don't think we should remove layout from these functions... they might come handy in the future.....

Reply:
Labelling this proposal as rejected.